### PR TITLE
Issue #2: Initial commit for SSL Certificate generation bundle

### PR DESCRIPTION
**Pull request by Stratehm**
Signed-off-by: Antoine Besnard <stratehm@hotmail.com>

### CODE REVIEW

**Comment by maggu2810** [comment]
Hello, I could have just a short look at this PR at the moment.

Why are you using an OSGi service instead of the bundle activator.
The question is more a technical one. Is it "okay" (conceptual) to use an OSGi service without providing something another service could ever consume? (I don't know.)

You are calling the generate function in the activate function. Shouldn't the activate return as soon as possible (to not blocking the OSGi service implementation) and delegate the generation to a task?

**Comment by kaikreuzer** [comment]
> The question is more a technical one. Is it "okay" (conceptual) to use an OSGi service without providing something another service could ever consume? (I don't know.)

In general, I would also expect such functionality in the Activator. The only reason might be, if one wants to defer the execution from bundle startup to DS, but I think DS blocks the startup in the same way, so there is not much benefit in doing so.

> You are calling the generate function in the activate function. Shouldn't the activate return as soon as possible (to not blocking the OSGi service implementation) and delegate the generation to a task?

Yes, it should return as fast as possible. But if we have to ensure that the certificate is created before the next start level is reached (and Jetty starts up), the fastest possible moment might be after the creation.
We should definitely check how much time this code takes on e.g. a RaspPi 1. If this exceeds 15 seconds, I would suggest that we think about an alternative solution.

**Comment by maggu2810** [comment]
>
> In general, I would also expect such functionality in the Activator. The
> only reason might be, if one wants to defer the execution from bundle
> startup to DS, but I think DS blocks the startup in the same way, so there
> is not much benefit in doing so.
>
> Just remember, if you would like to move this to the Activator, you have
to remove the lazy activation.

> Yes, it should return as fast as possible. But if we have to ensure that
> the certificate is created before the next start level is reached (and
> Jetty starts up), the fastest possible moment might be after the creation.
>
> Does jetty read the keystore file on startup only?
IMHO Relying on start level in OSGi is not the best approach.

> I would suggest that we think about an alternative solution.
>
If jetty reads the keystore only once, perhaps there is some other way to
force a reload.

**Comment by Stratehm** [comment]
Just for information, I have made this PR to be able discuss about the bundle on concrete bases.

> Why are you using an OSGi service instead of the bundle activator.

No special reason. I can use the bundle activator instead if it is a best practice.



>> You are calling the generate function in the activate function. Shouldn't the activate return as soon as possible (to not blocking the OSGi service implementation) and delegate the generation to a task?

>Yes, it should return as fast as possible. But if we have to ensure that the certificate is created before the next start level is reached (and Jetty starts up)

It is indeed to ensure the certificate generation happens before the Jetty startup.

> Does jetty read the keystore file on startup only?

It seems since there is an error at Jetty startup if it does not find the keystore.

> We should definitely check how much time this code takes on e.g. a RaspPi 1. If this exceeds 15 seconds, I would suggest that we think about an alternative solution.

I could try to start openHAB with the bundle on the first version of Rasp1 B, but I have a problem with the bundle logger when running the distrib:
```
SLF4J: The following loggers will not work because they were created
SLF4J: during the default configuration phase of the underlying logging system.
SLF4J: See also http://www.slf4j.org/codes.html#substituteLogger
SLF4J: org.openhab.io.certificate.generator.internal.CertificateGenerator
```

I do not know why.

**Comment by kaikreuzer** [comment]
> I do not know why.

It sounds as if your bundle is started before slf4j is fully initialized. Could you try to play around with the startlevels of the logback bundles to make sure that they start before your bundle?

**Comment by Stratehm** [comment]
On the Raspberry Pi B (256 MB), it takes from 20 seconds to 130 seconds (tried 6 times) to generate the 2048 bits keys. The generation time is random since it depends on the number of iterations needed to find the key prime numbers from random numbers.

=> Moving from RSA based certificate to Elliptic Curves. The generation time goes down to about 5 seconds for 256 bits keys (equivalent to 3072 bits keys with RSA). Moreover, with EC, the generation time is not random. 

I still have to find a way to generate the keys without JCE to allow the shade plugin to strip down the JAR.

**Comment by Stratehm** [comment]
The new implementation using 256-bit Elliptic Curve keys to generate the certificate is ready. It takes about 2 seconds to generate the whole certificate on the first version of Raspberry Pi B. The counterpart is that the bundle size has grown to 460 KB.

It now uses the BundleActivator instead of a Declarative Service.

**Comment by kaikreuzer** [comment]
Cool, thanks! Github claims that there are some merge conflicts, could you have a look and resolve them?

**Comment by Stratehm** [comment]
Done and squashed the commits.

File: bundles/io/pom.xml [hunk 1]
```diff
@@ -18,6 +18,7 @@
   <modules>
     <module>org.openhab.io.transport.serial</module>
     <module>org.openhab.io.jetty</module>
+	<module>org.openhab.io.certificate.generator</module>
```
Comment (maggu2810) [inline bundles/io/pom.xml]: It seems, that currently spaces are used to indent.
Could you change your tab to spaces?

File: products/org.openhab.product.runtime/org.openhab.runtime.product.product [hunk 2]
```diff
@@ -112,9 +112,11 @@
       <plugin id="org.openhab.core.compat1x" autoStart="true" startLevel="0" />
       <plugin id="org.openhab.core.init" autoStart="true" startLevel="1" />
       <plugin id="org.openhab.io.rest.docs" autoStart="true" startLevel="0" />
+      <plugin id="org.openhab.io.certificate.generator" autoStart="true" startLevel="2" />
       <plugin id="org.openhab.ui.classicui" autoStart="true" startLevel="0" />
       <plugin id="org.openhab.ui.dashboard" autoStart="true" startLevel="0" />
       <plugin id="org.openhab.ui.iconset.classic" autoStart="true" startLevel="0" />
+      <plugin id="org.openhab.ui.iconset.classic" autoStart="true" startLevel="0" />
```
Comment (maggu2810) [inline products/org.openhab.product.runtime/org.openhab.runtime.product.product]: Duplicate lines

File: distribution/BuildDistribution.launch [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
 <booleanAttribute key="M2_DEBUG_OUTPUT" value="false"/>
-<stringAttribute key="M2_GOALS" value="clean package"/>
+<stringAttribute key="M2_GOALS" value="clean install"/>
 <booleanAttribute key="M2_NON_RECURSIVE" value="false"/>
 <booleanAttribute key="M2_OFFLINE" value="false"/>
 <stringAttribute key="M2_PROFILES" value=""/>
 <listAttribute key="M2_PROPERTIES"/>
 <stringAttribute key="M2_RUNTIME" value="EMBEDDED"/>
-<booleanAttribute key="M2_SKIP_TESTS" value="false"/>
+<booleanAttribute key="M2_SKIP_TESTS" value="true"/>
```
Comment (maggu2810) [inline distribution/BuildDistribution.launch]: Why has the skip tests option been changed?

File: distribution/BuildDistribution.launch [hunk 4]
```diff
@@ -1,13 +1,13 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <launchConfiguration type="org.eclipse.m2e.Maven2LaunchConfigurationType">
 <booleanAttribute key="M2_DEBUG_OUTPUT" value="false"/>
-<stringAttribute key="M2_GOALS" value="clean package"/>
+<stringAttribute key="M2_GOALS" value="clean install"/>
```
Comment (maggu2810) [inline distribution/BuildDistribution.launch]: Why do we need to add the build product to the local repository now?

**Comment by maggu2810** [comment]
```java
    private static final String KEYSTORE_RELATIVE_PATH = "runtime/etc/keystore";
```

@kaikreuzer
Should we add the system properties for openhab (openhab.runtime, openhab.conf, openhab.userdata) to the standard distribution in the master branch, too?

**Comment by kaikreuzer** [comment]
From within Java code, the ConfigConstants class should normally be used. Maybe we can add a method getRuntimeFolder() to it? The code here itself should not depend on any system properties.

**Comment by maggu2810** [comment]
> The code here itself should not depend on any system properties.

Correct, this was not my intention, that someone uses the system properties directly.
I just asked, if we should add it. We should provide a utility class to get such paths.

> From within Java code, the ConfigConstants class should normally be used.

The ConfigConstants is part of ESH and not OH2.
Don't know, if this is a necessary constant of ESH based systems.

The OH2 jetty bundle (org.openhab.io.jetty) currently constructs the keystore (+truststore) path using:
system property "jetty.base" (default=".") + / + system property "jetty.keystore" ("etc/keystore")

This is IMHO openHAB 2 related only.
Perhaps we should simply add a system property jetty.keystore.path and add an OH2 ConfigConstants class, that offers a function getJettyKeyStorePath().

File: distribution/BuildDistribution.launch (hunk 3 above)
Comment (Stratehm) [inline distribution/BuildDistribution.launch]: It is a mistake, nothing to do in the commit.

File: distribution/BuildDistribution.launch (hunk 4 above)
Comment (Stratehm) [inline distribution/BuildDistribution.launch]: Idem as above. I will fix the comments. Thanks for the review.

Comment (maggu2810) [inline distribution/BuildDistribution.launch]: Thanks for the code. :)

**Comment by kaikreuzer** [comment]
> The OH2 jetty bundle (org.openhab.io.jetty) currently constructs the keystore (+truststore) path using:
system property "jetty.base" (default=".") + / + system property "jetty.keystore" ("etc/keystore")

So "runtime" is actually not required in the path at all, because the "." is already within this directory? Why does the constant above require the "runtime" folder then?

> Perhaps we should simply add a system property jetty.keystore.path and add an OH2 ConfigConstants class, that offers a function getJettyKeyStorePath().

As nobody else is supposed to use this, I would even say that we should not have a ConfigConstants Java class at all here, but simply use the system property jetty.keystore.path, since the Jetty home bundle uses it as well already.

**Comment by maggu2810** [comment]
> So "runtime" is actually not required in the path at all, because the "." is already within this directory? Why does the constant above require the "runtime" folder then?

Because I copied the lines using my working copy (karaf branch)...
The master branch is using:
```java
 <Set name="KeyStorePath"><SystemProperty name="jetty.base" default="." />/<SystemProperty name="jetty.keystore" default="runtime/etc/keystore"/></Set>
```

> but simply use the system property jetty.keystore.path, since the Jetty home bundle uses it as well already

The system property jetty.keystore.path does not exist at the moment.
The Jetty bundle does (IMHO) currently using:
```java
System.getProperty("jetty.base", ".") + "/" + System.getProperty("jetty.keystore", "runtime/etc/keystore");
```

Before we start adding this construct to another code line (hard coded default values in two different files), I would prefer to introduce such a new system property (e.g. jetty.keystore.path), and change the jetty-ssl.xml.

**Comment by kaikreuzer** [comment]
> I would prefer to introduce such a new system property (e.g. jetty.keystore.path), and change the jetty-ssl.xml.

I am ok with that. @Stratehm, Could you adapt this in your code and at the same time in the jetty-ssl.xml as well as the start scripts (adding the new system parameter there)?

**Comment by Stratehm** [comment]
It's done.

**Comment by kaikreuzer** [comment]
Wow, you are quick, thanks! I will have a look today and test it!

File: bundles/io/org.openhab.io.certificate.generator/src/main/java/org/openhab/io/certificate/generator/internal/CertificateGenerator.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    private void generateCertificate(KeyStore keystore) throws CertificateException, KeyStoreException {
+        try {
+            long startTime = System.currentTimeMillis();
+            org.bouncycastle.jce.spec.ECParameterSpec ecSpec = ECNamedCurveTable.getParameterSpec(CURVE_NAME);
+            ECField field = new ECFieldFp(ecSpec.getCurve().getField().getCharacteristic());
+            EllipticCurve curve = new EllipticCurve(field, ecSpec.getCurve().getA().toBigInteger(),
+                    ecSpec.getCurve().getB().toBigInteger());
+            ECPoint pointG = new ECPoint(ecSpec.getG().getXCoord().toBigInteger(),
+                    ecSpec.getG().getYCoord().toBigInteger());
+            ECParameterSpec spec = new ECParameterSpec(curve, pointG, ecSpec.getN(), ecSpec.getH().intValueExact());
```
Comment (kaikreuzer) [inline bundles/io/org.openhab.io.certificate.generator/src/main/java/org/openhab/io/certificate/generator/internal/CertificateGenerator.java]: This method does not exist in Java 1.7, see https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#intValueExact

**Comment by kaikreuzer** [comment]
Besides the dependency on Java 8, the code looks pretty nice!

One issue that now occurs is that the bundle starts before the logging is fully initialized:
```
SLF4J: The following loggers will not work because they were created
SLF4J: during the default configuration phase of the underlying logging system.
SLF4J: See also http://www.slf4j.org/codes.html#substituteLogger
SLF4J: org.openhab.io.certificate.generator.internal.CertificateGenerator
```
I have tried to add startlevel=1 to org.slf4j.api bundle, but this did not make any difference :-(
It is not really a problem of this PR, but we should try to find a solution for this.

One more wish: As the bundle contains 3rd party libs that are under a different license, it would be nice if you could add an about.html as described in https://eclipse.org/legal/epl/about.php. Thanks!

**Comment by Stratehm** [comment]
I have asked some help to license@eclipse.org since this bundle uses cryptography. I will update the about.html file later if needed.

> I have tried to add startlevel=1 to org.slf4j.api bundle, but this did not make any difference :-(
It is not really a problem of this PR, but we should try to find a solution for this.

It works with the org.openhab.io.certificate.generator bundle start @3 (if @2, the problem is still here) and org.eclipse.jetty.osgi.boot to start @4.

**Comment by kaikreuzer** [comment]
> I have asked some help to license@eclipse.org since this bundle uses cryptography.

We are not contributing this code to Eclipse, so there isn't really a reason to ask there for assistance :-) I mainly linked to that page in order to provide a template for the about.html.

**Comment by kaikreuzer** [comment]
> org.eclipse.jetty.osgi.boot to start @4.

But 4 is already the default start level - I am not sure if Jetty shouldn't come early. We can test and if it is a problem, we could set the default to 5 then.

**Comment by Stratehm** [comment]
> We are not contributing this code to Eclipse, so there isn't really a reason to ask there for assistance :-) I mainly linked to that page in order to provide a template for the about.html.

Indeed...let's see what they answer, it may still be interesting :)

> But 4 is already the default start level - I am not sure if Jetty shouldn't come early. We can test and if it is a problem, we could set the default to 5 then.

When running in Dev in Eclipse, the level is set to 2 by default for this bundle. 

Just for my info, where are defined the start levels of bundles for the packaged runtime (since they  are different from the ones defined in openHAB_Runtime.launch)

**Comment by kaikreuzer** [comment]
> Just for my info, where are defined the start levels of bundles for the packaged runtime

Right here: https://github.com/openhab/openhab2/blob/master/products/org.openhab.product.runtime/org.openhab.runtime.product.product#L38

**Comment by Stratehm** [comment]
Thanks. I had seen this file but I do not understand why the start levels defined in the file /runtime/server/configuration/config.ini in the package are not the same.

For example, for the org.eclipse.jetty.osgi.boot bundle.
In https://github.com/openhab/openhab2/blob/master/products/org.openhab.product.runtime/org.openhab.runtime.product.product#L38, the start level is 0: 
\<plugin id="org.eclipse.jetty.osgi.boot" autoStart="true" startLevel="0" /\>

And in runtime/server/configuration/config.ini of the package, the start level is 4: reference\:file\:org.eclipse.jetty.osgi.boot_9.2.12.v20150709.jar@4\:start

**Comment by kaikreuzer** [comment]
> I had seen this file but I do not understand why the start levels defined in the file /runtime/server/configuration/config.ini in the package are not the same.

I honestly have no clue myself - I would have expected them to be the same...

**Comment by kaikreuzer** [comment]
I have just re-tested it; if we change the start level of the generator to 3, everything works fine, so I guess we should do so.

One last remark: I am not too happy with the bundle name. I would prefer something like "org.openhab.io.certificategenerator" or "org.openhab.io.jetty.certificate". What do you think?

**Comment by Stratehm** [comment]
> Start level 0 in the product file means "default" and the default is 4. Therefore the generated config.ini file contains a 4.

Thanks for the explanations.

> One last remark: I am not too happy with the bundle name. I would prefer something like "org.openhab.io.certificategenerator" or "org.openhab.io.jetty.certificate". What do you think?

org.openhab.io.jetty.certificate sounds good.

> I have just re-tested it; if we change the start level of the generator to 3, everything works fine, so I guess we should do so.

Great. So, I will modify the start level and the bundle name, squash the commits and here we go :)

**Comment by kaikreuzer** [comment]
Sounds good!

**Comment by kaikreuzer** [comment]
@belovictor As you have entered https://github.com/openhab/openhab2/issues/2: Does this solution meet your expectation?

**Comment by kaikreuzer** [comment]
> org.openhab.io.jetty.certificate sounds good.

So why did you change it to org.openhab.io.certificate then?

**Comment by Stratehm** [comment]
Ouch...here is what happens when you are disturbed by your children...let's try again ;)

File: bundles/io/org.openhab.io.jetty.certificate/.classpath [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="src" path="src/main/java"/>
+	<classpathentry exported="true" kind="lib" path="lib/bcprov-jdk15on-1.52.jar" sourcepath="C:/Users/Win7Dev64/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.52/bcprov-jdk15on-1.52-sources.jar"/>
```
Comment (kaikreuzer) [inline bundles/io/org.openhab.io.jetty.certificate/.classpath]: you probably want to remove your sourcepath here!

File: bundles/io/org.openhab.io.jetty.certificate/about.html [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"/>
+<title>About</title>
+</head>
+<body lang="EN-US">
+<h2>About This Content</h2>
+ 
+<p><em>October 9, 2015</em></p>    
+<h3>License</h3>
+
+<p>The Eclipse Foundation makes available all content in this plug-in (&quot;Content&quot;).  Unless otherwise
```
Comment (kaikreuzer) [inline bundles/io/org.openhab.io.jetty.certificate/about.html]: Just noticed it now: Replace "The Eclipse Foundation" with "openHAB" here.

File: bundles/io/org.openhab.io.jetty.certificate/about.html [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+<p><em>October 9, 2015</em></p>    
+<h3>License</h3>
+
+<p>The Eclipse Foundation makes available all content in this plug-in (&quot;Content&quot;).  Unless otherwise 
+indicated below, the Content is provided to you under the terms and conditions of the
+Eclipse Public License Version 1.0 (&quot;EPL&quot;).  A copy of the EPL is available 
+at <a href="http://www.eclipse.org/legal/epl-v10.html">http://www.eclipse.org/legal/epl-v10.html</a>.
+For purposes of the EPL, &quot;Program&quot; will mean the Content.</p>
+
+<p>If you did not receive this Content directly from the Eclipse Foundation, the Content is
```
Comment (kaikreuzer) [inline bundles/io/org.openhab.io.jetty.certificate/about.html]: here as well

File: bundles/io/org.openhab.io.jetty.certificate/src/main/java/org/openhab/io/jetty/certificate/internal/CertificateGenerator.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
+package org.openhab.io.jetty.certificate.internal;
```
Comment (kaikreuzer) [inline bundles/io/org.openhab.io.jetty.certificate/src/main/java/org/openhab/io/jetty/certificate/internal/CertificateGenerator.java]: could you add the default license header here?

**Comment by kaikreuzer** [comment]
Oops, sorry for the last minute review - but I just found some small things to correct.
Now I will shut up, promised!!!

File: bundles/io/org.openhab.io.jetty.certificate/src/main/java/org/openhab/io/jetty/certificate/internal/CertificateGenerator.java (hunk 9 above)
Comment (Stratehm) [inline bundles/io/org.openhab.io.jetty.certificate/src/main/java/org/openhab/io/jetty/certificate/internal/CertificateGenerator.java]: This one ? 
Bundle-License: http://www.eclipse.org/legal/epl-v10.html;description=Eclipse Public License Version 1.0;link=http://www.eclipse.org/legal/eplfaq.php

Comment (kaikreuzer) [inline bundles/io/org.openhab.io.jetty.certificate/src/main/java/org/openhab/io/jetty/certificate/internal/CertificateGenerator.java]: No, the one from all other source files: https://github.com/openhab/openhab2/blob/master/bundles/core/org.openhab.core/src/main/java/org/openhab/core/OpenHAB.java#L1

Comment (Stratehm) [inline bundles/io/org.openhab.io.jetty.certificate/src/main/java/org/openhab/io/jetty/certificate/internal/CertificateGenerator.java]: Of course...really need to go to bed...

**Comment by Stratehm** [comment]
> Oops, sorry for the last minute review - but I just found some small things to correct.
Now I will shut up, promised!!!

No problem.

**Comment by kaikreuzer** [comment]
Thanks a lot for all your efforts! Let's merge!

**Pull request merged**